docs/architecture/adr-058-authz-hooks.md

```diff
@@ -0,0 +1,141 @@
+# ADR 058: Authz hooks
+
+## Changelog
+
+* 2022-08-02: Initial Draft
+
+## Status
+
+DRAFT Not Implemented
+
+## Abstract
+
+In order to extend the use cases covered by the `x/authz` module, we propose to implement a hook system so that external modules can perform stateful checks and/or state changes while handling a `MsgExec`. 
+
+## Context
+
+Currently, when handling `MsgExec` it is not possible to execute stateful checks before accepting an authorization, nor it is possible to perform state changes after it has been accepted. This makes it impossible to correctly implement some use cases like fee-based authorizations (as described inside [\#11583](https://github.com/cosmos/cosmos-sdk/issues/11583)).
+
+## Decision
+
+In order to improve the number of use cases that can be handled though the usage of `x/authz`, we proposed to implement a hook system within the authz keeper, similar to what other modules already use. 
+
+Particularly, we have thought about the following hook structure: 
+
+```go
+type AuthzHooks interface {
+    AfterAuthorizationAccepted(ctx sdk.Context, grantee sdk.AccAddress, authorization Authorization) error
+}
+```
+
+This can then be used as follows within the `x/auth` keeper structure: 
+
+```go
+type Keeper struct {
+    hooks authz.AuthzHooks
+    // ...
+}
```

Comment:
My first impression in looking at this ADR is that it's more intuitive to have `AfterAuthorizationAccepted` on the `Authorization` interface instead of the keeper.

I guess the challenge is: what's the function signature of `AfterAuthorizationAccepted`, and how to access the keepers, is that right?

Reply:
Yes, that's right. I've thought about it as well, but I ended up going with the hooks-based pattern since it's already implemented in other modules and might be easier to understand for developers without having the need to introduce a new one

Reply:
One idea could be:

```go
type Authorization interface {
  // ...same methods as before

  AfterAccepted(ctx sdk.Context) []sdk.Msg
}
```

`DispatchActions` will run the authz Msg, then must run _all_ the Msgs from `AfterAccepted` (can be nil array obviously) for the state to be commited. In your `FeeAuthorization` example, the return value would be a single `MsgSend`.

What do you think? (or other people too)

Reply:
I see. From an API perspective, it seems like if we wanted to reject an authorization based on stateful checks, it should happen in the `Accept` method, not in `AfterAuthorizationAccepted`. The after-accept keeper hook seems like a hack for rejection, it's also IMO confusing mentally as then each Authorization is not "standalone" anymore, but depends on the keepers implementations in which it's injected.

Reply:
BTW, what's your use case for stateful checks before rejection?

Reply:
> I see. From an API perspective, it seems like if we wanted to reject an authorization based on stateful checks, it should happen in the `Accept` method, not in `AfterAuthorizationAccepted`. The after-accept keeper hook seems like a hack for rejection, it's also IMO confusing mentally as then each Authorization is not "standalone" anymore, but depends on the keepers implementations in which it's injected.

I think this can then be solved by changing the hook name to something more easily understandable as `CheckAuthorization` or something else in that direction.


> BTW, what's your use case for stateful checks before rejection?

One use case that comes to the top of mind mind right now might be pre-given authorizations that need to be executable only when a certain condition happens. For example, I might want to give a user the authorization to perform a specific action, but only when they stake enough tokens or perform any kind of other activity. 

To do this currently you need to have an external observer that continuously checks for that condition and issues/revokes the authorization accordingly.

Instead of doing this, with stateful checks authorizations I could grant him an authorization that when executed checks for such condition and gets executed only if such condition is true (i.e. the user has staked enough tokens).

Reply:
I do agree that on `Accept` is cleaner to do these stateful checks, but I don't know enough about authz to know if that's possible?

Is the idea/solution here to just change the hook name?

Reply:
I also want to keep backwards-compatibility as much as possible, so if possible we should keep `Accept` as-is, and think about a separate mechanism to do stateful checks and state transitions.